[GraphTrace] include input tensor in memory usage

gfengTT opened the pull request:
Peak memory usage is calculated from the outer trace, so that it also includes input tensors.

### Ticket
Required by https://github.com/tenstorrent/tt-mlir/issues/4143

### Problem description
Avoid recalculating input tensor size in tt-mlir.

### What's changed
Also added a missing field in ConstraintQueryResponse.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

gfengTT committed:
[GraphTrace] include input tensor in memory usage

Peak memory usage is calculated from the outer trace, so that it also
includes input tensors.

gfengTT commented:
@rpavlovicTT in this way, input tensor size will be accounted for in op_trace

@mbezuljTT it looks like you originally wrote this code. Is it okay if we switch from inner trace to outer trace?

gfengTT commented:
@ayerofieiev-tt approval? :)

rpavlovicTT commented on ttnn/api/ttnn/graph/graph_query_op_constraints.hpp:
```diff
@@ -122,14 +122,13 @@ auto query_op_constraints(Op op, tt::tt_metal::distributed::MeshDevice* device,
             log_debug(tt::LogOp, "Error during graph capture: {}", e.what());
             return ConstraintQueryResponse{
-                ExecutionStatus::Error, {0, 0, 0}, /* output_tensor_spec= */ std::nullopt, e.what()};
+                ExecutionStatus::Error, {0, 0, 0, 0}, /* output_tensor_spec= */ std::nullopt, e.what()};
```
Annotate 0s

gfengTT replied:
Done

rpavlovicTT reviewed (changes requested):
@gfengTT can you add a few tests? For example, conv2d with `deactivate_activation=ON` and without. I'd expect some difference there. Also, go through existing unit tests and refurbish them if needed.

gfengTT committed:
Update test case to cover input tensor size

With input tensor in L1, the conv2d test will report an L1 peak of
190,572 and total peak of 391,400. After including input tensor size,
the result will be L1 peak 247,916 and total peak 448,744. I updated the
threshold so that the test only passes with input tensor size included.

gfengTT committed:
annotate values with the field names

gfengTT committed:
test conv2d with multiple configs

1. null config
2. deallocate_activation=true

gfengTT commented:
@rpavlovicTT I've made the conv2d test to try with both null config (`/0`) and deallocate_activation=true (`/1`). The total peak is:

Before this change (w/o input size):
`/0`: 391400
`/1`: 340200
After this change (w/ input size):
`/0`: 442600
`/1`: 391400

gfengTT committed:
Update tests/ttnn/unit_tests/gtests/test_graph_query_op_constraints.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
